README.md
# WP-IPIP

**WP-IPIP** 是一款根据 IP 显示评论来源地址的插件。插件使用 IPIP.net 免费版数据库。

## 安装

### 手动安装

本插件未包含在 WordPress Plugin 仓库中，所以需要手动安装。

将文件复制到 WordPress 的插件目录下即可，文件应该在 `/wp-content/plugins/wp-ipip` 目录下。

结构：

```
└── wp-content
    └── plugins
        └── wp-ipip
            ├── 17mon
            │   └── php
            │       ├── 17monipdb.datx
            │       ├── IP4datx.class.php
            │       └── readme.txt
            ├── Changelog.md
            ├── js
            │   └── wp-ipip.js
            ├── README.md
            ├── wp-ipip-options.html
            ├── wp-ipip-options.php
            └── wp-ipip.php
```

### IPIP数据库安装和更新

IPIP 数据库文件路径：`{插件路径}/17mon/php/17monipdb.datx`，直接用新的数据库覆盖此文件即可。插件包中目前自带一份中文版数据库。

你可以在IPIP的官网下载最新版的数据库：<https://www.ipip.net/download.html>

**注意：根据 IPIP.net 2018 年 3 月 发布的数据库信息，未来的格式支持中不再支持 DAT 格式。**

**所以从 0.2.0 版本开始，插件将仅支持 DATX 版本的数据库。对于英文版和一些有特殊情况而使用 DAT 格式的用户，请继续使用 0.1.X 版本。**

### 更新

插件未在 WordPress 上保存任何数据，所以建议更新时删除旧插件并上传新文件。

### 卸载

插件未在 WordPress 上保存任何数据，所以可以直接删除。

## 使用

插件启用后，将会在评论页面将会在 IP 下显示物理地址。

IPIP 免费版精度支持到地级市。

本插件不支持 IPIP 收费版数据库。

## 注意事项

插件使用的数据库文件需要在运行时载入内存，整个插件会占用约 4M 左右的内存。按照经验，Wordpress 在运行时要占用 28M 左右的内存。如果您的服务器对 PHP 运行内存限制为 32M 或者更低的话，会出现评论页面无法正确载入，或载入后页面功能失效的问题。

请先确认自己的 PHP 运行环境是否有类似限制。

## 已知兼容性问题

### PHP

至少 PHP5.3 以上版本

### akisment

与 akisment 兼容

### wp-useragent

与插件 wp-useragent 曾存在兼容性问题。具体原因是因为 wp-useragent 并未使用 Wordpress 推荐的 filter & return 方式输出结果，并非本插件的问题。已在此版本中优化。

## 链接

IPIP官方解析代码-PHP：<https://github.com/17mon/php>


Changelog.md
# Changelog

## 0.1.3

- Update IPIP database from 201707 to 201803.

I don't know why the database file is smaller than the older one.

IPIP.net will not support .dat database any longer, so a new version of WP-IPIP is on working for the .datx database.

## 0.1.2

- Update to new IPIP lib and database (201707).

Notice the database file is much larger than before, so it will cost more memory.

## 0.1.1

- 增加选项配置页面，提供部分安装信息
- 增加 Readme.md 中的信息
- 优化代码

## 0.1

- 首次提交


wp-ipip.php
<?php

/*
Plugin Name: WP-IPIP
Plugin URI: http://URI_Of_Page_Describing_Plugin_and_Updates
Description: 根据评论IP地址定位物理地址，使用IPIP.net数据库。
Version: 0.2.0
Author: Catscarlet
Author URI: https://www.catscarlet.com
*/

if (!function_exists('add_action')) {
    exit;
}

if (is_admin()) {
    add_action('admin_enqueue_scripts', 'load_wp_ipip_resources');
}

add_action('admin_menu', 'wpipip_add_option_page');

function wpipip_add_option_page()
{
    add_options_page('WP-IPIP 设置页面', 'WP-IPIP 设置', 'manage_options', 'wp-ipip/wp-ipip-options.php');
}

function load_wp_ipip_resources($hook_suffix)
{
    if ('edit-comments.php' == $hook_suffix) {
        wp_register_script('wp-ipip.js', plugin_dir_url(__FILE__).'js/wp-ipip.js', array('jquery'));
        wp_enqueue_script('wp-ipip.js');
        include_once WP_PLUGIN_DIR.'/wp-ipip/17mon/php/IP4datx.class.php';
        add_filter('comment_text', 'wp_ipip', 10, 2);
    }
}

function wp_ipip($comment_text, $comment = null)
{
    if (!$comment) {
        return $comment_text;
    }

    try {
        $results = IP::find($comment->comment_author_IP);
    } catch (Exception $e) {
        $location = 'WP-IPIP Caught exception: '.$e->getMessage();
        $wpipip_e = '<div class="wp-ipip-comment" display="none" id="wp-ipip-perfix-'.$comment->comment_ID.'" style="color:red">'.$location.'</div>';

        echo $wpipip_e;

        return $comment_text;
    }

    if ('N/A' == $results) {
        return $comment_text;
    }

    $location = array();
    foreach ($results as $str) {
        if (null != $str) {
            $location[] = $str;
        }
    }
    $info = implode(',', $location);

    $wpipip = '<div class="wp-ipip-comment" display="none" id="wp-ipip-perfix-'.$comment->comment_ID.'">地址: '.$info.'</div>';

    echo $wpipip;

    return $comment_text;
}


17mon/php/IP.class.php
<?php

/*
    全球 IPv4 地址归属地数据库(IPIP.NET 版)
    高春辉(pAUL gAO) <gaochunhui@gmail.com>
    Build 20170905 版权所有 IPIP.NET
    (C) 2006 - 2017 保留所有权利，北京天特信科技有限公司
    本代码仅用于 DAT 格式，请注意及时更新 IP 数据库版本
    数据问题请加 QQ 交流群: 346280296
    Code for PHP 5.3+ only!
*/

class IP
{
    private static $ip     = NULL;

    private static $fp     = NULL;
    private static $offset = NULL;
    private static $index  = NULL;

    public static function find($ip)
    {
        if (empty($ip) === TRUE)
        {
            return 'N/A';
        }

        $nip   = gethostbyname($ip);
        $ipdot = explode('.', $nip);

        if ($ipdot[0] < 0 || $ipdot[0] > 255 || count($ipdot) !== 4)
        {
            return 'N/A';
        }

        if (self::$fp === NULL)
        {
            self::init();
        }

        $nip2 = pack('N', ip2long($nip));

        $tmp_offset = (int)$ipdot[0] * 4;
        $start      = unpack('Vlen', self::$index[$tmp_offset] . self::$index[$tmp_offset + 1] . self::$index[$tmp_offset + 2] . self::$index[$tmp_offset + 3]);

        $index_offset = $index_length = NULL;
        $max_comp_len = self::$offset['len'] - 1024 - 4;
        for ($start = $start['len'] * 8 + 1024; $start < $max_comp_len; $start += 8)
        {
            if (self::$index{$start} . self::$index{$start + 1} . self::$index{$start + 2} . self::$index{$start + 3} >= $nip2)
            {
                $index_offset = unpack('Vlen', self::$index{$start + 4} . self::$index{$start + 5} . self::$index{$start + 6} . "\x0");
                $index_length = unpack('Clen', self::$index{$start + 7});

                break;
            }
        }

        if ($index_offset === NULL)
        {
            return 'N/A';
        }

        fseek(self::$fp, self::$offset['len'] + $index_offset['len'] - 1024);

        return explode("\t", fread(self::$fp, $index_length['len']));
    }

    private static function init()
    {
        if (self::$fp === NULL)
        {
            self::$ip = new self();

            self::$fp = fopen(__DIR__ . '/17monipdb.dat', 'rb');
            if (self::$fp === FALSE)
            {
                throw new Exception('Invalid 17monipdb.dat file!');
            }

            self::$offset = unpack('Nlen', fread(self::$fp, 4));
            if (self::$offset['len'] < 4)
            {
                throw new Exception('Invalid 17monipdb.dat file!');
            }

            self::$index = fread(self::$fp, self::$offset['len'] - 4);
        }
    }

    public function __destruct()
    {
        if (self::$fp !== NULL)
        {
            fclose(self::$fp);

            self::$fp = NULL;
        }
    }
}

?>

17mon/php/IP4datx.class.php
<?php

/*
    全球 IPv4 地址归属地数据库(IPIP.NET 版)
    高春辉(pAUL gAO) <gaochunhui@gmail.com>
    Build 20170905 版权所有 IPIP.NET
    (C) 2006 - 2017 保留所有权利，北京天特信科技有限公司
    本代码仅用于 DATX 格式，请注意及时更新 IP 数据库版本
    数据问题请加 QQ 交流群: 346280296
    Code for PHP 5.3+ only!
*/

class IP
{
    private static $ip     = NULL;

    private static $fp     = NULL;
    private static $offset = NULL;
    private static $index  = NULL;

    public static function find($ip)
    {
        if (empty($ip) === TRUE)
        {
            return 'N/A';
        }

        $nip   = gethostbyname($ip);
        $ipdot = explode('.', $nip);

        if ($ipdot[0] < 0 || $ipdot[0] > 255 || count($ipdot) !== 4)
        {
            return 'N/A';
        }

        if (self::$fp === NULL)
        {
            self::init();
        }

        $nip2 = pack('N', ip2long($nip));

        $tmp_offset = ((int)$ipdot[0] * 256 + (int)$ipdot[1]) * 4;
        $start      = unpack('Vlen', self::$index[$tmp_offset] . self::$index[$tmp_offset + 1] . self::$index[$tmp_offset + 2] . self::$index[$tmp_offset + 3]);

        $index_offset = $index_length = NULL;
        $max_comp_len = self::$offset['len'] - 262144 - 4;
        for ($start = $start['len'] * 9 + 262144; $start < $max_comp_len; $start += 9)
        {
            if (self::$index{$start} . self::$index{$start + 1} . self::$index{$start + 2} . self::$index{$start + 3} >= $nip2)
            {
                $index_offset = unpack('Vlen', self::$index{$start + 4} . self::$index{$start + 5} . self::$index{$start + 6} . "\x0");
                $index_length = unpack('nlen', self::$index{$start + 7} . self::$index{$start + 8});

                break;
            }
        }

        if ($index_offset === NULL)
        {
            return 'N/A';
        }

        fseek(self::$fp, self::$offset['len'] + $index_offset['len'] - 262144);

        return explode("\t", fread(self::$fp, $index_length['len']));
    }

    private static function init()
    {
        if (self::$fp === NULL)
        {
            self::$ip = new self();

            self::$fp = fopen(__DIR__ . '/17monipdb.datx', 'rb');
            if (self::$fp === FALSE)
            {
                throw new Exception('Invalid 17monipdb.datx file!');
            }

            self::$offset = unpack('Nlen', fread(self::$fp, 4));
            if (self::$offset['len'] < 4)
            {
                throw new Exception('Invalid 17monipdb.datx file!');
            }

            self::$index = fread(self::$fp, self::$offset['len'] - 4);
        }
    }

    public function __destruct()
    {
        if (self::$fp !== NULL)
        {
            fclose(self::$fp);

            self::$fp = NULL;
        }
    }
}

?>